File: functions/wib_week.js
/**
 * WIB (Asia/Jakarta, UTC+7, no daylight saving) — anchored ISO-8601-
 * style week id, for the weekly Top Global competition period.
 *
 * **Deliberately a new, separate module — `award_top_coins.js`'s own
 * `isoWeekId` is NOT modified or reused for this.** `isoWeekId`
 * computes week boundaries from raw UTC calendar dates; "Monday 00:00
 * WIB" is "Sunday 17:00 UTC the day before" — a 7-hour shift from
 * where `isoWeekId`'s own Monday-anchored math would place a UTC week
 * boundary. Reusing it unmodified would silently produce WIB-week
 * boundaries wrong by up to 7 hours. Confirmed via `grep -rln
 * "isoWeekId" functions/*.js` (excluding tests) that only
 * `award_top_coins.js` itself uses it — RISK-3's subscription backstop
 * uses no week-id concept at all (plain daily/weekly cron schedules) —
 * so `isoWeekId` is kept completely unmodified; nothing depends on
 * `wibWeekId` existing alongside it.
 *
 * The fixed `+7h` shift is exact and permanent here specifically
 * *because* WIB has no DST (Indonesia does not observe daylight
 * saving) — this trick is unsafe for a DST-observing zone and should
 * not be copied elsewhere without re-checking that assumption.
 */

/** WIB is a fixed UTC+7 offset, year-round — no DST to account for. */
const WIB_OFFSET_MS = 7 * 60 * 60 * 1000;

/**
 * ISO-week-numbered id (`YYYY-Www`) for the WIB week containing [date]
 * — the exact same string format `award_top_coins.js`'s `isoWeekId`
 * already uses, for consistency with the existing `weeklyCoinAwards/
 * {id}` convention, computed via the identical, already-proven
 * Monday-Thursday ISO-8601 anchor algorithm, just applied to the
 * WIB-shifted calendar date rather than the raw UTC one.
 *
 * The period boundary is a half-open interval: `[Monday 00:00:00.000
 * WIB, next Monday 00:00:00.000 WIB)` — the boundary instant itself
 * belongs to the period that is *starting*, not the one ending. This
 * falls out of the date-math itself (at exactly midnight WIB, the
 * shifted calendar date has already rolled to Monday) — no
 * special-casing needed.
 *
 * @param {Date} date the authoritative instant to resolve — callers
 *   must pass a server-derived timestamp (a Cloud Function trigger's
 *   own `event.time`, or an equivalent server clock read), never a
 *   client-supplied one. This function itself has no way to enforce
 *   that — it is a pure date-math utility — the caller's own choice of
 *   what `date` to pass is what makes the period identity
 *   server-authoritative or not.
 * @return {string} e.g. `"2026-W35"`
 */
function wibWeekId(date) {
  const shifted = new Date(date.getTime() + WIB_OFFSET_MS);
  const d = new Date(Date.UTC(
      shifted.getUTCFullYear(), shifted.getUTCMonth(), shifted.getUTCDate(),
  ));
  const dayNum = d.getUTCDay() || 7;
  d.setUTCDate(d.getUTCDate() + 4 - dayNum);
  const yearStart = new Date(Date.UTC(d.getUTCFullYear(), 0, 1));
  const weekNum = Math.ceil((((d - yearStart) / 86400000) + 1) / 7);
  return `${d.getUTCFullYear()}-W${String(weekNum).padStart(2, "0")}`;
}

/**
 * The start instant (Monday 00:00:00.000 WIB, expressed as its real
 * UTC instant) of the WIB week containing [date] — used by the payout
 * job to derive period boundaries without re-deriving them from a
 * periodId string.
 *
 * @param {Date} date
 * @return {Date}
 */
function wibWeekStart(date) {
  const shifted = new Date(date.getTime() + WIB_OFFSET_MS);
  const dayNum = shifted.getUTCDay() || 7;
  const mondayShifted = new Date(Date.UTC(
      shifted.getUTCFullYear(), shifted.getUTCMonth(),
      shifted.getUTCDate() - (dayNum - 1),
  ));
  return new Date(mondayShifted.getTime() - WIB_OFFSET_MS);
}

module.exports = {WIB_OFFSET_MS, wibWeekId, wibWeekStart};

File: lib/core/utils/wib_week.dart
/// WIB (Asia/Jakarta, UTC+7, no daylight saving) — anchored ISO-8601-
/// style week id, for the Weekly Global Ranking competition period.
///
/// **This is a client-side MIRROR of `functions/wib_week.js`'s
/// `wibWeekId`, used ONLY to know which already-written, read-only
/// `globalScorePeriods/{periodId}/users` document to fetch — it never
/// influences scoring or ranking itself.** The server (Cloud Functions,
/// via the CloudEvent's own `event.time`) is the sole authority on which
/// period an attempt's points actually land in; this function exists
/// purely so the app can compute "what period should I be reading right
/// now" without an extra round-trip. If this function's date math ever
/// drifted from the server's own `wibWeekId`, the practical effect would
/// be the client reading the WRONG (empty or stale) period document —
/// a display bug, not a security issue, since the client has no write
/// path to any of this data at all (`firestore.rules` denies every
/// client write to `globalScorePeriods`/`globalScorePeriodAwards`
/// unconditionally). Kept in exact lock-step with the server
/// implementation regardless — see `wib_week_test.dart`, which mirrors
/// `functions/wib_week.test.js`'s own boundary cases line for line.
///
/// The fixed `+7h` shift is exact and permanent specifically *because*
/// WIB has no DST (Indonesia does not observe daylight saving) — unsafe
/// to copy for a DST-observing zone without re-checking that assumption.
library;

/// WIB is a fixed UTC+7 offset, year-round — no DST to account for.
const wibOffset = Duration(hours: 7);

/// ISO-week-numbered id (`YYYY-Www`) for the WIB week containing [utc] —
/// the exact same string format the server's own `award_top_coins.js`
/// (`isoWeekId`, historical) and `wib_week.js` (`wibWeekId`, current)
/// use.
///
/// The period boundary is a half-open interval: `[Monday 00:00:00.000
/// WIB, next Monday 00:00:00.000 WIB)` — the boundary instant itself
/// belongs to the period that is *starting*, not the one ending. This
/// falls out of the date-math itself (at exactly midnight WIB, the
/// shifted calendar date has already rolled to Monday) — no special-
/// casing needed, mirroring `functions/wib_week.js`'s own `wibWeekId`.
///
/// @param utc must be a UTC [DateTime] (`DateTime.now().toUtc()`, or any
///   already-UTC instant) — passing a local-time [DateTime] would shift
///   by the device's own timezone on top of the explicit WIB shift below,
///   silently double-applying an offset.
String wibWeekId(DateTime utc) {
  assert(utc.isUtc, 'wibWeekId requires a UTC DateTime — call .toUtc() first');
  final shifted = utc.add(wibOffset);
  var d = DateTime.utc(shifted.year, shifted.month, shifted.day);
  final dayNum = d.weekday; // Dart's DateTime.weekday is already 1=Mon..7=Sun
  d = d.add(Duration(days: 4 - dayNum));
  final yearStart = DateTime.utc(d.year, 1, 1);
  final weekNum = (((d.difference(yearStart).inDays) + 1) / 7).ceil();
  return '${d.year}-W${weekNum.toString().padLeft(2, '0')}';
}

/// The start instant (Monday 00:00:00.000 WIB, expressed as its real UTC
/// instant) of the WIB week containing [utc] — mirrors
/// `functions/wib_week.js`'s own `wibWeekStart`.
DateTime wibWeekStart(DateTime utc) {
  assert(
    utc.isUtc,
    'wibWeekStart requires a UTC DateTime — call .toUtc() first',
  );
  final shifted = utc.add(wibOffset);
  final dayNum = shifted.weekday;
  final mondayShifted = DateTime.utc(
    shifted.year,
    shifted.month,
    shifted.day - (dayNum - 1),
  );
  return mondayShifted.subtract(wibOffset);
}

/// The current active WIB week's period id, for "what period should the
/// app be reading right now" — a thin, obviously-named wrapper around
/// [wibWeekId] so call sites read intent-first rather than repeating
/// `wibWeekId(DateTime.now().toUtc())` at every call site.
String currentWibPeriodId() => wibWeekId(DateTime.now().toUtc());
